### PR TITLE
`Development`: Migrate client code for pipes and loading indicator components

### PR DESCRIPTION
**Pull request by PaRangger**
#### General

- [x] I tested **all** changes and their related features with **all** corresponding user types on a test server.
- [x] This is a small issue that I tested locally and was confirmed by another developer on a test server.
- [x] Language: I followed the [guidelines for inclusive, diversity-sensitive, and appreciative language](https://docs.artemis.cit.tum.de/dev/guidelines/language-guidelines/).
- [x] I chose a title conforming to the [naming conventions for pull requests](https://docs.artemis.cit.tum.de/dev/development-process/development-process.html#naming-conventions-for-github-pull-requests).

#### Client
- [x] **Important**: I implemented the changes with a very good performance, prevented too many (unnecessary) REST calls and made sure the UI is responsive, even with large data (e.g. using paging).
- [x] I **strictly** followed the [client coding and design guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/client/).
- [x] I added multiple integration tests (Jest) related to the features (with a high test coverage), while following the [test guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/client-tests/).

### Motivation and Context

This is an effort to migrate some core components to the new angular apis. The process was covered in the all hands meeting and can be seen [here](https://confluence.ase.in.tum.de/spaces/ArTEMiS/pages/236914354/Core+Client+Migration+%E2%86%92+Needs+help+team). This PR addresses some core components which are used frequently in the communication module:

| ID | File | Standalone | Signals | Inject |
|-----|-------|-----------------:|------------|---------:|
| 387 | /shared/loading-indicator-container/loading-indicator-container.component.html | ✅ | ❌ | ✅ |
| 388 | /shared/loading-indicator-container/loading-indicator-container.component.ts | ❌ | ❌ | ✅ |
| 416 | /shared/pipes/html-for-markdown.pipe.ts | ❌ | ✅ | ❌ |
| 417 | /shared/pipes/html-for-posting-markdown.pipe.ts | ❌ | ✅ | ❌ |
| 410 | /shared/pipes/artemis-translate.pipe.ts | ❌ | ✅ | ❌ |

### Description

I replaced `@Input`, `@Output`, `@ViewChilds` with their corresponding signal equivalent and made sure to call them as functions in their templates. I replaced constructors with inject() functions. I converted them to `standalone: true` components. Lastly, I tried to fix the failing tests.

### Steps for Testing

Prerequisites:
- 1 Artemis User

**These pipes and components are used frequently throughout artemis. Therefore just observe if any problems occur on the entire platform. In these steps I will just guide you to a single occurence of the corresponding pipe/component.**

Loading Indicator
1. Log into Artemis
2. Navigate to course with communication enabled
3. Go to the communication tab
4. Select a channel
5. For a brief moment you should be able to see a loading indicator
6. Afterwards it should display the channel messages

HtmlMarkdownPipe/PostingMarkdownPipe
1. Log into Artemis
2. Navigate to course with communication enabled
3. Go to the communication tab
7. Write a post in any channel, make sure to format it using the editor
8. Check if it is rendered properly

Translate Pipe
1. Log in to Artemis
2. Navigate to course with communication enabled
3. Go to the communication tab
9. In the "All messages" tab, or when searching for a term, hover over the "sorting" arrow
10. The tooltip should be shown appropriately
11. Change the language on the top right
12. Hover again over the arrow, the tooltip should have changed to the accoriding language

[](https://artemis-test1.artemis.cit.tum.de)
[](https://artemis-test2.artemis.cit.tum.de)
[](https://artemis-test3.artemis.cit.tum.de)
[](https://artemis-test4.artemis.cit.tum.de)
[](https://artemis-test5.artemis.cit.tum.de)
[](https://artemis-test6.artemis.cit.tum.de)
[](https://artemis-test9.artemis.cit.tum.de)

#### Code Review
- [x] Code Review 1
- [x] Code Review 2
#### Manual Tests
- [x] Test 1
- [x] Test 2

#### Client

| Class/File | Line Coverage | Confirmation (assert/expect) |
|------------|--------------:|---------------------:|
| loading-indicator-container.component.ts | 75% | ✅ |
| artemis-translate.pipe.ts | 100% | ✅ |
| html-for-markdown.pipe.ts | 100% | ✅ |
| html-for-posting-markdown.pipe.ts | 100% | ✅ |

### Screenshots

Loading Indicator
Bildschirmfoto 2024-11-27 um 14 57 40

HtmlMarkdownPipe/PostingMarkdownPipe
Bildschirmfoto 2024-11-27 um 14 56 40

Translate Pipe
Bildschirmfoto 2024-11-27 um 14 55 32

## Release Notes

- **New Features**
	- Enhanced loading indicator functionality to display based on dynamic loading state.
	- Added new pipes for handling HTML in markdown content, allowing for more flexible content rendering.

- **Bug Fixes**
	- Improved test setups for various components to ensure accurate testing of functionality.

- **Documentation**
	- Updated module structures to reflect changes in pipe and component declarations.

- **Tests**
	- Streamlined test configurations to utilize mock implementations for better isolation and clarity.

### CODE REVIEW

**Review by HawKhiem — DISMISSED** [review]
Tested on TS5. Works as described

Screenshot 2024-11-27 195214
Screenshot 2024-11-27 195228

File: src/main/webapp/app/shared/pipes/artemis-translate.pipe.ts [hunk 1]
```diff
@@ -4,16 +4,16 @@ import { TranslatePipe, TranslateService } from '@ngx-translate/core';
...
  */
 export class ArtemisTranslatePipe implements PipeTransform, OnDestroy {
     private translatePipe: TranslatePipe;
-    constructor(
-        private translateService: TranslateService,
-        private changeDetectorRef: ChangeDetectorRef,
-    ) {
+
+    // We cannot inject() ChangeDetectorRef here
+    constructor(translateService: TranslateService, changeDetectorRef: ChangeDetectorRef) {
```
Comment (SimonEntholzer) [inline src/main/webapp/app/shared/pipes/artemis-translate.pipe.ts:16]: You could improve the comment here, by explaining _why_ we cannot do that here (or maybe even find a work around to make it work?)

**Review by SimonEntholzer — CHANGES_REQUESTED** [review]
Nice changes, I added one small comment

File: src/main/webapp/app/shared/pipes/artemis-translate.pipe.ts (hunk 1 above)
Comment (PaRangger) [reply src/main/webapp/app/shared/pipes/artemis-translate.pipe.ts]: Thanks for the feedback! 😄 I tried to extend the comment in this [commit](https://github.com/ls1intum/Artemis/pull/9880/commits/2925c0d5cd9b9aaf86f23beac3775e7949541e69). Unfortunately, I already tried some ways to get around it, but I could not find a good solution. 😅  But I am open for suggestions, if anyone could think of something.

**Review by sachmii — APPROVED** [review]
Tested on TS1, everything works as described.

Bildschirmfoto 2024-11-28 um 09 11 42
Bildschirmfoto 2024-11-28 um 09 14 16
Bildschirmfoto 2024-11-28 um 09 14 07

**Review by cremertim — APPROVED** [review]
code changes lgtm

**Review by SimonEntholzer — APPROVED** [review]
Code

**Review by HawKhiem — APPROVED** [review]
LGTM. Reapprove

**Review by FelixTJDietrich — APPROVED** [review]
Code looks good, and client tests are passing :) 
Tested on TS1 works as expected. Thanks for helping with the migration.

**Pull request merged**